<!--{
	"Title": "Defer、Panic和Recover",
	"Template": true
}-->

<!--{
	"Title": "Defer, Panic, and Recover",
	"Template": true
}-->

<div class="english">
<p>
Go has the usual mechanisms for control flow: if, for, switch, goto.  It also
has the go statement to run code in a separate goroutine.  Here I'd like to
discuss some of the less common ones: defer, panic, and recover.
</p>
</div>

<p>
Go语言提供一般的流程控制语句：if、for、switch、goto。
同时它还提供go语句来执行一个Go程。这里我们将介绍几个不太常见的语句：defer、panic和recover。
</p>

<div class="english">
<p>
A <b>defer statement</b> pushes a function call onto a list. The list of saved
calls is executed after the surrounding function returns. Defer is commonly
used to simplify functions that perform various clean-up actions.
</p>
</div>

<p>
一个<b>defer</b>调用的函数将被暂时保存到调用列表中。保存的调用列表在当前环境返回的时候被执行。
Defer一般可以用于简化代码，执行各种清理操作。
</p>

<div class="english">
<p>
For example, let's look at a function that opens two files and copies the
contents of one file to the other:
</p>
</div>

<p>
让我们演示一个文件复制的例子：函数需要打开两个文件，然后将其中一个文件的内容复制到另一个文件。
</p>

{{code "/doc/progs/defer.go" `/func CopyFile/` `/STOP/`}}

<div class="english">
<p>
This works, but there is a bug. If the call to os.Create fails, the
function will return without closing the source file. This can be easily
remedied by putting a call to src.Close before the second return statement,
but if the function were more complex the problem might not be so easily
noticed and resolved. By introducing defer statements we can ensure that the
files are always closed:
</p>
</div>

<p>
上面的代码虽然能够工作，但是隐藏一个bug。
如果第二个os.Open调用失败，那么会在没有释放source文件资源的情况下返回。
虽然我们可以通过在第二个返回语句前添加src.Close()调用来修复这个bug；
但是当代码变得复杂时，类似bug将很难被发现和修复。
通过defer语句，我们可以确保每个文件被关闭：
</p>

{{code "/doc/progs/defer2.go" `/func CopyFile/` `/STOP/`}}

<div class="english">
<p>
Defer statements allow us to think about closing each file right after opening
it, guaranteeing that, regardless of the number of return statements in the
function, the files <i>will</i> be closed.
</p>
</div>

<p>
Defer语言可以让我们在打开文件时就思考如何关闭文件。
不管函数如何返回，文件关闭语句始终会被执行。
</p>

<div class="english">
<p>
The behavior of defer statements is straightforward and predictable. There are
three simple rules:
</p>
</div>

<p>
Defer语句的行为简单且可预测。
有三个基本原则：
</p>

<div class="english">
<p>
1. <i>A deferred function's arguments are evaluated when the defer statement is
evaluated.</i>
</p>
</div>

<p>
1. <i>当defer调用函数的时候，函数用到的每个参数和变量的值也会被计算。</i>
</p>

<div class="english">
<p>
In this example, the expression "i" is evaluated when the Println call is
deferred. The deferred call will print "0" after the function returns.
</p>
</div>

{{code "/doc/progs/defer.go" `/func a/` `/STOP/`}}

<div class="english">
<p>
2. <i>Deferred function calls are executed in Last In First Out order
</i>after<i> the surrounding function returns.</i>
</p>
</div>

<p>
2. <i>Defer调用的函数将在当前函数返回的时候，以后进先出的顺序执行。</i>
</p>

<div class="english">
<p>
This function prints "3210":
</p>
</div>

{{code "/doc/progs/defer.go" `/func b/` `/STOP/`}}

<div class="english">
<p>
3. <i>Deferred functions may read and assign to the returning function's named
return values.</i>
</p>
</div>

<p>
3. <i>Defer调用的函数可以在返回语句执行后读取或修改命名的返回值。</i>
</p>

<div class="english">
<p>
In this example, a deferred function increments the return value i <i>after</i>
the surrounding function returns. Thus, this function returns 2:
</p>
</div>

<p>
在这个例子中，defer语句将会在当前函数返回后将i增加1。实际上，函数会返回2：
</p>

{{code "/doc/progs/defer.go" `/func c/` `/STOP/`}}

<div class="english">
<p>
This is convenient for modifying the error return value of a function; we will
see an example of this shortly.
</p>
</div>

<p>
利用该特性，我们可以方便修改函数的错误返回值；以后应该可以看到类似的例子。
</p>

<div class="english">
<p>
<b>Panic</b> is a built-in function that stops the ordinary flow of control and
begins <i>panicking</i>. When the function F calls panic, execution of F stops,
any deferred functions in F are executed normally, and then F returns to its
caller. To the caller, F then behaves like a call to panic. The process
continues up the stack until all functions in the current goroutine have
returned, at which point the program crashes. Panics can be initiated by
invoking panic directly. They can also be caused by runtime errors, such as
out-of-bounds array accesses.
</p>
</div>

<p>
<b>Panic</b>是一个内置的函数：停止当前控制流，然后开始<i>panicking</i>。
当F函数调用panic，F函数将停止执行后续的普通语句，但是之前的defered函数调用仍然被正常执行，然后再返回到F的调用者。
对于F函数的调用者，F 的行为和直接调用panic函数类似。
以上的处理流程会一直沿着调用栈回朔，直到当前的goroutine返回引起程序崩溃！
Panics可以通过直接调用panic方式触发，也可以由某些运行时错误触发，例如：数组的越界访问。
</p>

<div class="english">
<p>
<b>Recover</b> is a built-in function that regains control of a panicking
goroutine. Recover is only useful inside deferred functions. During normal
execution, a call to recover will return nil and have no other effect. If the
current goroutine is panicking, a call to recover will capture the value given
to panic and resume normal execution.
</p>
</div>

<p>
<b>Recover</b>也是一个内置函数:用于从 panicking 恢复。
Recover 和 defer 配合使用会非常有用。
对于一个普通的执行流程，调用recover将返回nil，也没有任何效果。
但如果当前goroutine处于panicking状态，recover调用会捕获触发panic时的参数，并且恢复到正常的执行流程。
</p>

<div class="english">
<p>
Here's an example program that demonstrates the mechanics of panic and defer:
</p>
</div>

<p>
下面的例子演示了panic 和 defer配合使用的技术：
</p>

{{code "/doc/progs/defer2.go" `/package main/` `/STOP/`}}

<div class="english">
<p>
The function g takes the int i, and panics if i is greater than 3, or else it
calls itself with the argument i+1. The function f defers a function that calls
recover and prints the recovered value (if it is non-nil). Try to picture what
the output of this program might be before reading on.
</p>
</div>

<p>
函数g有一个整型参数i，在参数i大于3时将触发panics异常，否则将以i+1为参数递归调用自己。
函数f通过defers中调用recover来捕获异常，并输出触发异常的参数（如果不是nil的话）。
在查看输出结果前，读者可以自己现预测一下输出结果。
</p>

<div class="english">
<p>
The program will output:
</p>
</div>

<p>
程序的输出：
</p>

<pre>Calling g.
Printing in g 0
Printing in g 1
Printing in g 2
Printing in g 3
Panicking!
Defer in g 3
Defer in g 2
Defer in g 1
Defer in g 0
Recovered in f 4
Returned normally from f.</pre>

<div class="english">
<p>
If we remove the deferred function from f the panic is not recovered and
reaches the top of the goroutine's call stack, terminating the program. This
modified program will output:
</p>
</div>

<p>
如果我们从函数f中移除deferred语句，panic在扩散到goroutine栈顶前将不会被捕获，最终会引起 程序崩溃。
下面是修改后的输出结果：
</p>

<pre>Calling g.
Printing in g 0
Printing in g 1
Printing in g 2
Printing in g 3
Panicking!
Defer in g 3
Defer in g 2
Defer in g 1
Defer in g 0
panic: 4

panic PC=0x2a9cd8
[stack trace omitted]</pre>

<div class="english">
<p>
For a real-world example of <b>panic</b> and <b>recover</b>, see the
<a href="/pkg/encoding/json/">json package</a> from the Go standard library.
It decodes JSON-encoded data with a set of recursive functions.
When malformed JSON is encountered, the parser calls panic to unwind the
stack to the top-level function call, which recovers from the panic and returns
an appropriate error value (see the 'error' and 'unmarshal' methods of
the decodeState type in
<a href="/src/pkg/encoding/json/decode.go">decode.go</a>).
</p>
</div>

<p>
一个真实的<b>panic</b>和<b>recover</b>配合使用的用例可以参考标准库：
<a href="/pkg/encoding/json/">json package</a>。
它提供JSON格式的解码，当遇到非法格式的输入时会抛出panic异常，然后panicking扩散到上一级调用者堆栈，
由上一级调用者通过recovers捕获panic和错误信息（参考
<a href="/src/pkg/encoding/json/decode.go">decode.go</a>中的‘error’和‘unmarshal’）。
</p>

<div class="english">
<p>
The convention in the Go libraries is that even when a package uses panic
internally, its external API still presents explicit error return values.
</p>
</div>

<p>
Go库的实现习惯：即使在pkg内部使用了panic，但是在导出API时会被转化为明确的错误值。
</p>

<div class="english">
<p>
Other uses of <b>defer</b> (beyond the file.Close example given earlier)
include releasing a mutex:
</p>
</div>

<p>
另一个使用<b>defer</b>的场景是释放mutex（参考前面给出的file.Close()例子）：
</p>

<pre>mu.Lock()
defer mu.Unlock()</pre>

<div class="english">
<p>
printing a footer:
</p>
</div>

<p>
打印页眉和页脚：
</p>

<pre>printHeader()
defer printFooter()</pre>

<div class="english">
<p>
and more.
</p>
</div>

<p>
更多。
</p>

<div class="english">
<p>
In summary, the defer statement (with or without panic and recover) provides an
unusual and powerful mechanism for control flow.  It can be used to model a
number of features implemented by special-purpose structures in other
programming languages. Try it out.
</p>
</div>

<p>
总而言之，defer语句（不管是否包含panic和recover）提供了一种不同寻常且十分强大的控制流机制。
它可以用于模拟一些其他语言中的某些特殊的语法结构。享受defer带来的便利吧！
</p>
